Plugins/GitHub3/GitHubCredentialsPrompt.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using GitCommands.Utils;
+using GitUI;
+
+namespace GitHub3
+{
+    public partial class GitHubCredentialsPrompt : Form
+    {
+        private bool _require2ndFactorCode = false;
+
+        public GitHubCredentialsPrompt()
+        {
+            InitializeComponent();
+            lblError.Text = string.Empty;
+        }
+
+        private void btn_OK_Click(object sender, EventArgs e)
+        {
+            var tokenTask = ThreadHelper.JoinableTaskFactory.RunAsync(() => GenerateOAuthTokenAsync(txtGitHubLogin.Text, txtGitHubPassword.Text, txtSecondFactor.Text));
+            var token = tokenTask.Join();
+            if (!string.IsNullOrWhiteSpace(token))
+            {
+                GitHubLoginInfo.OAuthToken = token;
+
+                MessageBox.Show(this, "Successfully retrieved OAuth token.", "GitHub Authorization");
+                DialogResult = DialogResult.OK;
+                Close();
+            }
+        }
+
+        private void GitHubCredentialsPrompt_Shown(object sender, EventArgs e)
+        {
+            txtGitHubLogin.Focus();
+        }
+
+        private void ButtonCancel_Click(object sender, EventArgs e)
+        {
+            DialogResult = DialogResult.Cancel;
+            Close();
+        }
+
+        private void TxtGitHubLogin_TextChanged(object sender, EventArgs e)
+        {
+            UpdateOkButtonState();
+        }
+
+        private void UpdateOkButtonState()
+        {
+            btn_OK.Enabled = !string.IsNullOrWhiteSpace(txtGitHubLogin.Text) &&
+                             !string.IsNullOrWhiteSpace(txtGitHubPassword.Text) &&
+                             (!_require2ndFactorCode || string.IsNullOrWhiteSpace(txtSecondFactor.Text));
+        }
+
+        public class GitHubToken
+        {
+            public string token { get; set; }
+        }
+
+        private async Task<string> GenerateOAuthTokenAsync(string login, string password, string secondFactorOtp)
+        {
+            // https://developer.github.com/v3/auth/#using-the-oauth-authorizations-api-with-two-factor-authentication
+            const string otpHeaderKey = "X-GitHub-OTP";
+            const string githubScopesJson = "{\"scopes\":[\"repo\", \"public_repo\"],\"note\":\"Token for Git Extensions\"}";
+            using (var httpClient = new HttpClient())
+            {
+                using (var request = new HttpRequestMessage(new HttpMethod("POST"), "https://api.github.com/authorizations"))
```

Comment:
I thought about that for begin of the support of GitHub enterprise.

Reply:
Yesterday I visited my former colleagues and they use GHE. They'd love this

Reply:
>They'd love this

I hope they will be able to provide some time to test the build because I don't have access to GHE. 


In fact, I think that it is not too difficult to achieve once this PR is done.

The 2 things missing:

* add the URL in the settings (easy and I already done it, even if I think we must revamp the settings for a good UX)
* we need a method in the plugin interface that is called when the repository is changed to switch from GH to GHE (or the opposite) if needed depending on the repository settings.